docs/plugins/builtins.md

```diff
@@ -0,0 +1,631 @@
+# Builtin Plugins
+
+## _SecretGenerator_
+
+### Usage via `kustomization.yaml`
+
+#### field name: `secretGenerator`
+
+Each entry in the argument list
+results in the creation of
+one Secret resource
+(it's a generator of n secrets).
+
+```
+secretGenerator:
+- name: app-tls
+  files:
+  - secret/tls.cert
+  - secret/tls.key
+  type: "kubernetes.io/tls"
+- name: app-tls-namespaced
+  # you can define a namespace to generate
+  # a secret in, defaults to: "default"
+  namespace: apps
+  files:
+  - tls.crt=catsecret/tls.cert
+  - tls.key=secret/tls.key
+  type: "kubernetes.io/tls"
+- name: env_file_secret
+  envs:
+  - env.txt
+  type: Opaque
+```
+
+### Usage via plugin
+
+#### Arguments
+
+> ```
+> types.GeneratorOptions
+> types.SecretArgs
+> ```
+
+#### Example
+
+> ```
+> apiVersion: builtin
+> kind: SecretGenerator
+> metadata:
+>   name: mySecret
+>   namespace: whatever
+> behavior: merge
+> envs:
+> - a.env
+> - b.env
+> files:
+> - obscure=longsecret.txt
+> literals:
+> - FRUIT=apple
+> - VEGETABLE=carrot
+> ```
+
+## _ConfigMapGenerator_
+
+### Usage via `kustomization.yaml`
+
+#### field name: `configMapGenerator`
+
+Each entry in this list results in the creation of
+one ConfigMap resource (it's a generator of n maps).
+
+The example below creates two ConfigMaps. One with the
+names and contents of the given files, the other with
+key/value as data.
+
+Each configMapGenerator item accepts a parameter of
+`behavior: [create|replace|merge]`.
+This allows an overlay to modify or
+replace an existing configMap from the parent.
+
+```
+configMapGenerator:
+- name: myJavaServerProps
+  files:
+  - application.properties
+  - more.properties
+- name: myJavaServerEnvVars
+  literals:	
+  - JAVA_HOME=/opt/java/jdk
+  - JAVA_TOOL_OPTIONS=-agentlib:hprof
+```
+
+It is also possible to
+[define a key](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#define-the-key-to-use-when-creating-a-configmap-from-a-file)
+to set a name different than the filename.
+
+The example below creates a ConfigMap
+with the name of file as `myFileName.ini`
+while the _actual_ filename from which the
+configmap is created is `whatever.ini`.
+
+```
+configMapGenerator:
+- name: app-whatever
+  files:
+  - myFileName.ini=whatever.ini
+```
+
+### Usage via plugin
+#### Arguments
+> ```
+> types.GeneratorOptions
+> types.ConfigMapArgs
+> ```
+
+#### Example
+> ```
+> apiVersion: builtin
+> kind: ConfigMapGenerator
+> metadata:
+>   name: myMap
+> envs:
+> - devops.env
+> - uxteam.env
+> literals:
+> - FRUIT=apple
+> - VEGETABLE=carrot
+> ```
+
+
+
+
+## _NamespaceTransformer_
+### Usage via `kustomization.yaml`
+
+#### field name: `namespace`
+
+Adds namespace to all resources
+
+```
+namespace: my-namespace
+```
+
+### Usage via plugin
+#### Arguments
+> ```
+> FieldSpecs []config.FieldSpec
+> ```
+
+#### Example
+> ```
+> apiVersion: builtin
+>  kind: NamespaceTransformer
+>  metadata:
+>    name: notImportantHere
+>    namespace: test
```

Comment:
The syntax of the namespace transformer is kind of misleading. the "metadata/name" is not important but the "metadata/namespace" is. Unlike the other transformer (prefix, suffix), that field is not part of the "spec" of the Transformer but part of the metadata.

Reply:
Not every plugin has the config under `spec`. Even for prefix/suffix transformer, it's `prefix` and `suffix` instead of `spec`. The transformer/generator plugin read the whole k8s styled yaml as its configuration, not only `spec` field.

Reply:
Let me rephrase. All the other transformers have the addditional parameters at the top level....basically the fieldSpecs level, the metadata field is actually irrelevant most of the time.

```yaml
apiVersion: builtin
kind: LabelTransformer
metadata:
  name: notImportantHere
labels:
  app: myApp
  env: production
fieldSpecs:
  - path: metadata/labels
    create: true
```

or 

```yaml
apiVersion: builtin
kind: PrefixSuffixTransformer
metadata:
  name: notImportantHere
prefix: baked-
suffix: -pie
fieldSpecs:
  - path: metadata/name
```

But not for the namespace transformer. The consistent way of doing, should be like this:

```yaml
apiVersion: builtin
kind: NamespaceTransformer
metadata:
  name: notImportantHere
namespace: test
fieldSpecs:
  - path: metadata/namespace
```

The other issue is [here](https://github.com/kubernetes-sigs/kustomize/blob/master/pkg/target/kusttarget.go#L313). The name is actually important for ImageTransformer if you need multiple image transformations. The namespace is supposed to be metadata/namespace is supposed to be completely irrelevant but is not, and the namespace transformer ends up being in a different namespace than all the other transformers. This will be an issue when we will try to apply kustomize on top of the transformer config.yaml

```go
func (kt *KustTarget) configureExternalTransformers() ([]resmap.Transformer, error) {
	ra := accumulator.MakeEmptyAccumulator()
	err := kt.accumulateResources(ra, kt.kustomization.Transformers)
	if err != nil {
		return nil, err
	}
	return kt.pLdr.LoadTransformers(kt.ldr, ra.ResMap())
}```

Reply:
@jbrette yes, the namespace transformer is an outlier.
see comments here: https://github.com/kubernetes-sigs/kustomize/pull/1280
it's philosophically wrong, but in practice is works and allows for a smaller config file, fwiw.
we should probably 'fix' it if for no other reason than to avoid this kind of confusion/conversation.
Not in this PR though.

Reply:
Thank you for this clarification, and for this comment especially: https://github.com/kubernetes-sigs/kustomize/pull/1280#issuecomment-506904144!